File: backend/services/school_service.py
from typing import Dict, Any, List
from datetime import datetime, timezone
import logging
from services.supabase_client import get_supabase_client

logger = logging.getLogger(__name__)

class SchoolService:
    @staticmethod
    def get_schools() -> List[Dict[str, Any]]:
        try:
            supabase = get_supabase_client()
            resp = supabase.table("schools").select("*").order("created_at", desc=True).execute()
            return resp.data
        except Exception as e:
            logger.error(f"Failed to get schools: {e}")
            raise

    @staticmethod
    def get_school_by_id(school_id: str) -> Dict[str, Any]:
        try:
            supabase = get_supabase_client()
            resp = supabase.table("schools").select("*").eq("id", school_id).execute()
            if not resp.data:
                return None
            return resp.data[0]
        except Exception as e:
            logger.error(f"Failed to get school {school_id}: {e}")
            raise

    @staticmethod
    def create_school(data: Dict[str, Any]) -> Dict[str, Any]:
        try:
            supabase = get_supabase_client()
            resp = supabase.table("schools").insert(data).execute()
            return resp.data[0]
        except Exception as e:
            logger.error(f"Failed to create school: {e}")
            raise

    @staticmethod
    def update_school(school_id: str, data: Dict[str, Any]) -> Dict[str, Any]:
        try:
            supabase = get_supabase_client()
            data["updated_at"] = datetime.now(timezone.utc).isoformat()
            resp = supabase.table("schools").update(data).eq("id", school_id).execute()
            if not resp.data:
                return None
            return resp.data[0]
        except Exception as e:
            logger.error(f"Failed to update school {school_id}: {e}")
            raise

    @staticmethod
    def deactivate_school(school_id: str) -> Dict[str, Any]:
        try:
            supabase = get_supabase_client()
            data = {
                "status": "inactive",
                "updated_at": datetime.now(timezone.utc).isoformat()
            }
            resp = supabase.table("schools").update(data).eq("id", school_id).execute()
            if not resp.data:
                return None
            return resp.data[0]
        except Exception as e:
            logger.error(f"Failed to deactivate school {school_id}: {e}")
            raise

    @staticmethod
    def get_school_stats(school_id: str) -> Dict[str, Any]:
        # Currently a placeholder that returns 0. 
        # Future phases will run COUNT() queries on linked tables.
        try:
            # We would eventually query here:
            # supabase = get_supabase_client()
            # teachers_count = ...
            return {
                "teachers": 0,
                "students": 0,
                "parents": 0,
                "assessments": 0,
                "highRiskStudents": 0
            }
        except Exception as e:
            logger.error(f"Failed to get stats for school {school_id}: {e}")
            raise

File: src/services/school.service.ts
const API_URL = 'http://127.0.0.1:8000/api/schools';

export interface School {
  id: string;
  school_name: string;
  school_code: string;
  district?: string;
  address?: string;
  city?: string;
  state?: string;
  country?: string;
  postal_code?: string;
  phone?: string;
  email?: string;
  principal_name?: string;
  logo_url?: string;
  academic_year?: string;
  status: 'active' | 'inactive' | 'pending';
  created_at?: string;
  updated_at?: string;
}

export interface SchoolStats {
  teachers: number;
  students: number;
  parents: number;
  assessments: number;
  highRiskStudents: number;
}

export const schoolService = {
  async getSchools(): Promise<School[]> {
    const response = await fetch(API_URL);
    if (!response.ok) throw new Error('Failed to fetch schools');
    const data = await response.json();
    return data.schools;
  },

  async getSchoolById(id: string): Promise<School> {
    const response = await fetch(`${API_URL}/${id}`);
    if (!response.ok) throw new Error('Failed to fetch school details');
    const data = await response.json();
    return data.school;
  },

  async createSchool(schoolData: Partial<School>): Promise<School> {
    const response = await fetch(API_URL, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(schoolData)
    });
    const data = await response.json();
    if (!response.ok) throw new Error(data.detail || 'Failed to create school');
    return data.school;
  },

  async updateSchool(id: string, schoolData: Partial<School>): Promise<School> {
    const response = await fetch(`${API_URL}/${id}`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(schoolData)
    });
    const data = await response.json();
    if (!response.ok) throw new Error(data.detail || 'Failed to update school');
    return data.school;
  },

  async deactivateSchool(id: string): Promise<School> {
    const response = await fetch(`${API_URL}/${id}/status`, {
      method: 'PATCH'
    });
    if (!response.ok) throw new Error('Failed to deactivate school');
    const data = await response.json();
    return data.school;
  },

  async getSchoolStats(id: string): Promise<SchoolStats> {
    const response = await fetch(`${API_URL}/${id}/stats`);
    if (!response.ok) throw new Error('Failed to fetch school stats');
    const data = await response.json();
    return data.stats;
  }
};
